irohad/ordering/impl/ordering_gate_cache/on_demand_cache.cpp

```diff
@@ -13,24 +15,47 @@ using namespace iroha::ordering::cache;
 // TODO: IR-1864 13.11.18 kamilsa use nvi to separate business logic and locking
 // logic
 
-void OnDemandCache::addToBack(
+namespace {
+  uint64_t countTransactions(const OrderingGateCache::BatchesSetType &batches) {
+    return std::accumulate(batches.begin(),
+                           batches.end(),
+                           0u,
+                           [](const uint64_t txs_quantity, const auto &batch) {
+                             return txs_quantity
+                                 + boost::size(batch->transactions());
+                           });
+  }
+}  // namespace
+
+OnDemandCache::OnDemandCache(uint64_t max_cache_size)
+    : max_cache_size_(max_cache_size) {}
+
+bool OnDemandCache::addToBack(
     const OrderingGateCache::BatchesSetType &batches) {
   std::unique_lock<std::shared_timed_mutex> lock(mutex_);
-  circ_buffer.back().insert(batches.begin(), batches.end());
+  const uint64_t input_size = countTransactions(batches);
+
+  if (circ_buffer.back().first + input_size > max_cache_size_) {
+    return false;
+  }
+  circ_buffer.back().first += input_size;
+  circ_buffer.back().second.insert(batches.begin(), batches.end());
+  return true;
 }
 
 void OnDemandCache::remove(const OrderingGateCache::HashesSetType &hashes) {
   std::unique_lock<std::shared_timed_mutex> lock(mutex_);
   for (auto &batches : circ_buffer) {
-    for (auto it = batches.begin(); it != batches.end();) {
+    for (auto it = batches.second.begin(); it != batches.second.end();) {
       if (std::any_of(it->get()->transactions().begin(),
                       it->get()->transactions().end(),
                       [&hashes](const auto &tx) {
                         return hashes.find(tx->hash()) != hashes.end();
                       })) {
         // returns iterator following the last removed element
         // hence there is no increment in loop iteration_expression
-        it = batches.erase(it);
+        batches.first -= countTransactions(batches.second);
```

Comment:
It subtracts the count of transaction in all the batches, but only a single batch `it` is removed below. Should it be the following?
```suggestion
        batches.first -= boost::size((*it)->transactions());
```

Reply:
need to write a test